request.go

```diff
@@ -0,0 +1,251 @@
+package gomatrixserverlib
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"github.com/matrix-org/util"
+	"golang.org/x/crypto/ed25519"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// A MatrixRequest is a request to send to remote server or a request received
+// from a remote server.
+// Matrix requests are signed by building a JSON object and signing it
+type MatrixRequest struct {
+	// fields implement the JSON format needed for signing
+	// specified in https://matrix.org/docs/spec/server_server/unstable.html#request-authentication
+	fields struct {
+		Content     rawJSON                      `json:"content,omitempty"`
+		Destination string                       `json:"destination"`
+		Method      string                       `json:"method"`
+		Origin      string                       `json:"origin"`
+		RequestURI  string                       `json:"uri"`
+		Signatures  map[string]map[string]string `json:"signatures,omitempty"`
+	}
+}
+
+// NewMatrixRequest creates a matrix request. Takes an HTTP method, a
+// destination homeserver and a request path which can have a query string.
+func NewMatrixRequest(method, destination, requestURI string) MatrixRequest {
```

Comment:
Please give an example as there are a ton of subtle things which are missing from the comments like:
 - Is `method` case-sensitive?
 - Does `destination` need `https://`? Does it accept path segments? Port numbers?
 - Does `requestURI` need a leading `/`? Does it support fragments in addition to query parameters?